### PR TITLE
refactor(compiler-cli): use `ngDevMode` guard for `setClassMetadata` call.

### PR DESCRIPTION
**Pull request by JoostK**
Prior to this change, the `setClassMetadata` call would be invoked
inside of an IIFE that was marked as pure. This allows the call to be
tree-shaken away in production builds, as the `setClassMetadata` call
is only present to make the original class metadata available to the
testing infrastructure. The pure marker is problematic, though, as the
`setClassMetadata` call does in fact have the side-effect of assigning
the metadata into class properties. This has worked under the assumption
that only build optimization tools perform tree-shaking, however modern
bundlers are also able to elide calls that have been marked pure so this
assumption does no longer hold. Instead, an `ngDevMode` guard is used
which still allows the call to be elided but only by tooling that is
configured to consider `ngDevMode` as constant `false` value.

### CODE REVIEW

**Comment by JoostK** [comment]
I looked into the symbol diffs a bit and I suspect that the change to `ngDevMode` means that Rollup now elides the code in a different order, probably later in the pipeline compared to dropping the pure call outright (although this is an educated guess). This causes function inlining to result in slightly different code. The `getBindingsEnabled` symbol is somewhat more interesting, as it appears that Terser is no longer able to prove it is always true, whereas it was before. Configuring Terser to do 4 passes instead of 3 allows `appendChild` and `getNodeInjectable` to be inlined yet again, but `getBindingsEnabled` still remains.

Here's an archive of before and after files for `cyclic_import` and `hello_world` bundles for anyone who's interested: 
[ng39987.zip](https://github.com/angular/angular/files/5647481/ng39987.zip).

**Comment by JoostK** [comment]
A much bigger issue occurs for aio payload sizes:

```
FAIL: Commit 7c5c48a8eca067dae80f63306359e466c6423ab7 uncompressed main-es2015
      exceeded expected size by 500 bytes or >1% (expected: 447349, actual: 489043).
```

:-(

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 1]
```diff
@@ -181,7 +181,8 @@ runInEachFileSystem(() => {
       const jsContents = fs.readFile(_(`/node_modules/test-package/index.js`)).replace(/\s+/g, ' ');
       expect(jsContents)
           .toContain(
-              '/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FooDirective, ' +
+              '(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ' +
```
Comment (crisbeto) [inline packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:184]: Drive-by comment: won't this still leave the empty IIFE in place when the code behind the dev mode check is removed? 

We should be able to do `(typeof ngDevMode === "undefined" || ngDevMode) && (function () {...})()` instead which will drop the entire block.

Comment (JoostK) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:184]: Yeah I thought that it would have been sufficient to do `(typeof ngDevMode === 'undefined' || ngDevMode) && setClassMetadata(...);` and that's how I started in 7c5c48a8eca067dae80f63306359e466c6423ab7, but it resulted in all decorator symbols (e.g. `Directive`) to be retained, strangely enough. I don't know exactly why that was the case as the references themselves were correctly removed, but somehow they were no longer tree-shaken out. I also tried to still wrap the `setClassMetadata` call in an IIFE with the dev mode check outside, as you suggested, but IIRC that also didn't work. I even played around with an additional Terser pass but it didn't help.

The current pattern is also used for guarding the `setNgModuleScope` call with `ngJitMode` checks. It does behave well in terms of tree-shaken and the IIFE will also be elided, as optimizers are apparently clever enough. Note also that [the CLI has a dedicated transform](https://github.com/angular/angular-cli/blob/eb30a92e8ad4bf27461010b1917d1fcffd10ff31/packages/ngtools/webpack/src/transformers/remove-ivy-jit-support-calls.ts) to delete these calls even before they reach Terser, although I'm not sure what the effect would be without that transform.

Comment (crisbeto) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:184]: I see, that does seem weird. It's worth noting that my example is different from 7c5c48a8eca067dae80f63306359e466c6423ab7. Mine would be the equivalent of doing the following which still has the IIFE:

```
if (typeof ngDevMode === "undefined" || ngDevMode) {
  (function() {
    setClassMetadata(...);
  })();
}
```

Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:184]: Maybe the extra internal IIFE is enough to solve the problem by creating a new scope.  Perhaps without it the minimizers assume that the `Component` classes are still living in the main scope?

Comment (JoostK) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:184]: IIRC that didn't work:

> I also tried to still wrap the setClassMetadata call in an IIFE with the dev mode check outside, as you suggested, but IIRC that also didn't work. I even played around with an additional Terser pass but it didn't help.

Comment (crisbeto) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:184]: What about wrapping it in the dev mode check _and_ leaving the pure annotation in place?

Comment (JoostK) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:184]: The pure annotation should not be used, as it could incorrectly result in dropping code even when the metadata is needed. E.g. Webpack 5 is able to drop pure annotated statements during bundling, without the need for an optimizer.

Is there something about the current way that isn't desirable, that I may be missing?

**Review by petebacondarwin — APPROVED** [review]
Assuming it is not possible to make the generated output simpler, I am happy with this.
Thanks @JoostK

**Review by AndrewKushnir — APPROVED** [review]
LGTM, thanks @JoostK 👍 
It's great that the generated code is now consistent with the `ɵɵsetNgModuleScope` case!

**Review by clydin — APPROVED** [review]
LGTM

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:346409095:BASE:346409116:1607465961219:c6930b7b).

**Comment by AndrewKushnir** [comment]
FYI, presubmit is successful for the changes in this PR, but it looks like there are conflicts with master branch, so I'm adding the "cleanup" label for now. Thank you.

**Comment by JoostK** [comment]
merge-assistance: previous presubmit was green, the rebase was only resolving conflicts in the golden files which are not synced into g3, so I doubt that this needs another presubmit.
